source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,162 @@
-			this _mutex unlock()
-		} else {
+_Task: abstract class {
+	_mutexUpdateTime: static Int = 1
+	_state := _PromiseState Unfinished
+	_freeDirectly: Bool
+	_mutex: Mutex
+	mutex ::= this _mutex
+	run: abstract func (mutex: Mutex)
+	wait: func -> Bool {
+		status := false
+		while (!status) {
+			this _mutex lock()
+			if (this _state != _PromiseState Unfinished) {
+				status = (this _state == _PromiseState Finished)
+				this _mutex unlock()
+				break
+			}
 			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
-			super()
+			Time sleepMilli(This _mutexUpdateTime)
 		}
+		status
 	}
-	run: override func {
-		this _execute()
+	cancel: func -> Bool {
+		status := false
 		this _mutex lock()
-		this _finished = true
-		if (this _freeOnCompletion) {
-			this _mutex unlock()
-			this free()
-		} else {
-			this _mutex unlock()
-			this _finishedCondition broadcast()
+		if (this _state == _PromiseState Unfinished) {
+			this _state = _PromiseState Cancelled
+			status = true
 		}
+		this _mutex unlock()
+		status
 	}
-	wait: func {
+	_finishedTask: func {
 		this _mutex lock()
-		if (!this _finished)
-			this _finishedCondition wait(this _mutex)
+		if (this _state != _PromiseState Cancelled)
+			this _state = _PromiseState Finished
 		this _mutex unlock()
+		if (this _freeDirectly)
+			this free()
+	}
+}
+
+_ActionTask: class extends _Task {
+	_action: Func
+	init: func (=_action, =_mutex)
+	run: func (=_mutex) {
+		this _action()
+		this _finishedTask()
+	}
+}
+
+_ResultTask: class <T> extends _Task {
+	_result: Cell<T>
+	_action: Func -> T
+	init: func (=_action, =_mutex)
+	run: func (=_mutex) {
+		temporary := Cell<T> new(this _action())
+		this _result = temporary
+		this _finishedTask()
+	}
+}
+
+_TaskPromise: class extends Promise {
+	_task: _Task
+	init: func (=_task)
+	free: override func {
+		this _task free()
+		super()
+	}
+	wait: override func -> Bool {
+		this _task wait()
+	}
+	cancel: override func -> Bool {
+		//TODO: Interrupt executing thread and have it move on to the next task in queue
+		this _task cancel()
+	}
+}
+
+_TaskFuture: class <T> extends Future<T> {
+	_task: _ResultTask<T>
+	init: func (=_task)
+	free: override func {
+		this _task free()
+		super()
+	}
+	wait: override func -> Bool {
+		this _task wait()
+	}
+	wait: func ~default (defaultValue: T) -> T {
+		status := this wait()
+		status ? this _task _result[T] : defaultValue
+	}
+	getResult: func (defaultValue: T) -> T {
+		status := (this _task _state == _PromiseState Finished)
+		status ? this _task _result[T] : defaultValue
+	}
+	cancel: override func -> Bool {
+		//TODO: Interrupt executing thread and have it move on to the next task in queue
+		this _task cancel()
 	}
 }
 
 ThreadPool: class {
+	_globalMutex := Mutex new()
 	_threads: Thread[]
-	_jobs := BlockedQueue<ThreadJob> new()
-	_mutex := Mutex new()
-	_allFinishedCondition := WaitCondition new()
-	_activeJobs: Int
+	_threadMutexes: Mutex[]
```

Comment:
This _must_ be combined into a worker. This creates a lot of messy code, which may contain errors.